### PR TITLE
Parallel hyperopt

### PR DESCRIPTION
**Pull request by cjmcgill**
Added checkpoint functions hyperparameter optimization to allow hyperopt to 1) be restarted after failing 2) start again after normal termination but adding additional iterations, and 3) run with multiple instances in parallel adding checkpoints to the same directory. Also added a feature to change the number of random iterations before guided suggestions are supplied.

Potential to add an additional feature to make trial histories out of previous runs that were done outside of hyperopt, but that has more edge cases with weird behavior than these other features so am leaving this off for now.

### CODE REVIEW

**Comment by cjmcgill** [comment]
Just realized I didn't update the readme. Will do that.

**Comment by cjmcgill** [comment]
I went back and added the additional feature to load in past training results as hyperopt trials. Now someone can attempt what they think will be the best combination of parameters or lightly grid search the space and include it in the hyperparameter optimization history.

**Comment by hesther** [comment]
Am currently testing various different cases and noting issues as they come up. General thoughts:

- Maybe we could log more info to `verbose.log` and the terminal output. For example, how many previous iterations are read in from `hyperopt_checkpoint_dir` or manual_trial_dirs`, possibly also their parameters and scores. Messages just as “Starting random search” and “Starting focussed search via TPE” or something like that so that the user knows whether TPE is still in the random stage or already started doing a focussed search. 

- write checkpoints automatically (I would suggest to `--log_dir` directly) instead of having to specify `--hyperopt_checkpoint_dir` (and use that only for read not write purposes). Side effect: If one continues a search in a new directory (specified via `--log_dir`), otherwise `quiet.log` and `verbose.log` will be written to the new directory, but the pickle files to the old directory - I think it would be better to keep everything together (add logs and pkl to old folder), or completely separate (all new logs and pkls to new folder) if different read and write paths are specified.

- when running in parallel, the logs look messy because we write the parameters first, then run the trial, then the scores. So if machine one and two start a trial, they will write the parameter set, then one of them will finish first, log that and start a new one, then the other machine will finish, producing a log like: 
```
{'depth': 3, 'dropout': 0.4, 'ffn_num_layers': 2, 'hidden_size': 1300}  #this is the first trial, machine 1
{'depth': 4, 'dropout': 0.4, 'ffn_num_layers': 2, 'hidden_size': 800} #this is the second trial, machine 2
num params: 2,146,401
1.6457599433993608 +/- 0.0 rmse #this is the result of the first trial, machine 1 (might be mistaken as trial 2)
{'depth': 2, 'dropout': 0.25, 'ffn_num_layers': 2, 'hidden_size': 1400} #this is the third trial, machine 1
num params: 5,437,901
1.6808909465508415 +/- 0.0 rmse #this is the result of the second trial, machine 2 (might be mistaken as trial 3)
...
```
maybe we could first run the trial, and then write the parameters and scores at the same time to ensure they are always together.

- in sequential mode, once TPE enters into the focussed search, each new parameter set is chosen based on the previous results. E.g. if trials 1-10 already exist, the parameters for trial 11 are chosen based on trial 1-10. When running in parallel, every machine accesses the same previous checkpoints to come up with a new parameter set to try. So, machine 1 would conduct trial 11 based on trial 1-10 and machine 2 would conduct trial 12 based on trial 1-10 (not 1-11). Trial 11 and 12 will likely not have the same parameters because TPE has some randomness in how it chooses a new set (I think). So the sequential and parallel mode will naturally yield different outcomes. I think we should add one sentence to the README, saying that they are not fully equivalent. Furthermore, running in parallel leads to a larger number of iterations than specified in num_iters, because e.g. for num_iters=20 a new iteration can be started on multiple machines if 19 trials have finished, leading to num_iters+num_processes-1 trials (or less, depending on the timing of the individual processes). If we wanted to circumvent that, we would have to count the number of started processes in addition to the finished ones (but that is difficult to do).

Will add more comments once I test more functions

**Comment by hesther** [comment]
A general observation (I think not a bug): If I run hyperparameter optimization twice, I get the exact same result twice. If I restart after X epochs, I get different results for X+1, X+2. Is this a result of how TPE initialises its random seeds? Do we need to fix that? With the current behavior, a failed and restarted run would explore slightly different options than a run that finished without interruption.

File: chemprop/hyperopt_utils.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+    # Load hyperopt trials object from each file
+    loaded_trials = Trials()
+    if previous_trials is not None:
+        loaded_trials = merge_trials(loaded_trials,previous_trials.trials)
+
+    for path in hyperopt_checkpoint_files:
+        with open(path,'rb') as f:
+            trial = pickle.load(f)
+            loaded_trials = merge_trials(loaded_trials, trial.trials)
```
Comment (hesther) [inline chemprop/hyperopt_utils.py:64]: Why is the empty `Trials()` in `loaded_trials` merged in? I.e. I am not sure why line 58 and 59 is necessary. Wouldn't it be enough to have something like:
```
if previous_trials is not None:
    loaded_trials=previous_trials.trials
else:
    loaded_trials=Trials()
```

(maybe I am missing something, though)

File: chemprop/hyperparameter_optimization.py [hunk 2]
```diff
@@ -79,25 +78,53 @@ def objective(hyperparams: Dict[str, Union[int, float]]) -> float:
...
+        fmin(
+            objective,
+            SPACE,
+            algo=partial(tpe.suggest, n_startup_jobs=args.startup_random_iters),
+            max_evals=len(trials) + 1,
+            trials=trials,
+            rstate=np.random.RandomState(hyperopt_seed),
+        )
+
+        # Create a trials object with only the last instance by merging the last data with an empty trials object
```
Comment (hesther) [inline chemprop/hyperparameter_optimization.py:121]: I guess this is the same question as above, why is it necessary to merge something with an empty trials object? Since this is repeated here, I guess it is not a bug but somehow necessary, and I am curious why :)

File: chemprop/hyperopt_utils.py (hunk 1 above)
Comment (cjmcgill) [reply chemprop/hyperopt_utils.py:64]: So what I found is that when you merge data into a Trials object to make it a larger Trials object, that is an inplace operation. So if you didn't create a new blank Trials instance here to merge into, then anything you added to `previous_trials` would be added to the Trials object that you fed in as an argument to the function. This keeps it from modifying that object outside the function.

Since I want to load in checkpoints anew every time anew every time, I don't want to accumulate that data anywhere. I just want it in a new fresh object every time. And since the same Trials object is being fed in as previous_trials every iteration, it wouldn't just accumulate the loaded trials but would accumulate many duplicate copies of them, which would then be counted against the total number of iterations.

**Comment by cjmcgill** [comment]
> A general observation (I think not a bug): If I run hyperparameter optimization twice, I get the exact same result twice. If I restart after X epochs, I get different results for X+1, X+2. Is this a result of how TPE initialises its random seeds? Do we need to fix that? With the current behavior, a failed and restarted run would explore slightly different options than a run that finished without interruption.

If you clear out your checkpoints folder, it will give you the same behavior because it will use the same seeds. That's the best that I could figure for recreatability.

 When I set the seeds it's using at every iteration, they're always different and I'm logging them in the checkpoint directory so they're never reused. If seeds were ever reused, then you could get duplicate parameters from concurrently running parallel processes and you would also get duplicates from the random space search in the initial random trials.

File: chemprop/hyperparameter_optimization.py (hunk 2 above)
Comment (cjmcgill) [reply chemprop/hyperparameter_optimization.py:121]: This is a slightly different version than above. I can't slice a Trials object, only the Trials.trials list of dictionaries that make up its constituent data, and I need to extract the last one to save by itself every iteration of the hyperopt search. I can't figure out a way to take a dict from Trials.trials and make it back into a Trials object directly, I can only figure out how to merge that data dict in with an empty Trials(), which is what I'm doing with merge_trials. It's possible that there's a cleaner way to do this, but it requires a better understanding of hyperopt trials and this is not exactly a use case that's in their documentation.

Comment (cjmcgill) [reply chemprop/hyperparameter_optimization.py:121]: Hmmm. Though I actually haven't tried what would happen if I did `Trials(Trials.trials)`. That might create a new instance of Trials with the data. I will test this.

Comment (cjmcgill) [reply chemprop/hyperparameter_optimization.py:121]: Confirmed this does not work.

**Comment by cjmcgill** [comment]
> write checkpoints automatically (I would suggest to --log_dir directly) instead of having to specify --hyperopt_checkpoint_dir (and use that only for read not write purposes). Side effect: If one continues a search in a new directory (specified via --log_dir), otherwise quiet.log and verbose.log will be written to the new directory, but the pickle files to the old directory - I think it would be better to keep everything together (add logs and pkl to old folder), or completely separate (all new logs and pkls to new folder) if different read and write paths are specified.

The read and write paths for pkl files need to be the same place so that parallel instances can all deposit and draw from the same pool. You don't want instance A to write to dirA and instance B to read from dirA but write to dirB. Then you have one-way information where instance A only knows about its own trials but instance B knows about both trials. And then if you start an instance C, who does it draw from?

**Comment by cjmcgill** [comment]
> when running in parallel, the logs look messy because we write the parameters first, then run the trial, then the scores. So if machine one and two start a trial, they will write the parameter set, then one of them will finish first, log that and start a new one, then the other machine will finish, producing a log like:
maybe we could first run the trial, and then write the parameters and scores at the same time to ensure they are always together.

This seems like a good idea, but it gets messy again once we add logging descriptions at each iteration. You could get multiple entries that read "Starting iteration 5/20, random search strategy (5 remaining)" for each parallel instance. Which I guess is okay.

**Comment by hesther** [comment]
> > write checkpoints automatically (I would suggest to --log_dir directly) instead of having to specify --hyperopt_checkpoint_dir (and use that only for read not write purposes). Side effect: If one continues a search in a new directory (specified via --log_dir), otherwise quiet.log and verbose.log will be written to the new directory, but the pickle files to the old directory - I think it would be better to keep everything together (add logs and pkl to old folder), or completely separate (all new logs and pkls to new folder) if different read and write paths are specified.
> 
> The read and write paths for pkl files need to be the same place so that parallel instances can all deposit and draw from the same pool. You don't want instance A to write to dirA and instance B to read from dirA but write to dirB. Then you have one-way information where instance A only knows about its own trials but instance B knows about both trials. And then if you start an instance C, who does it read from?

I see, then I guess we should keep the two different paths but if `--hyperopt_checkpoint_dir` is not specified we could set it to `--log_dir`

**Comment by hesther** [comment]
> > when running in parallel, the logs look messy because we write the parameters first, then run the trial, then the scores. So if machine one and two start a trial, they will write the parameter set, then one of them will finish first, log that and start a new one, then the other machine will finish, producing a log like:
> > maybe we could first run the trial, and then write the parameters and scores at the same time to ensure they are always together.
> 
> This seems like a good idea, but it gets messy again once we add logging descriptions at each iteration. You could get multiple entries that read "Starting iteration 5/20, random search strategy (5 remaining)" for each parallel instance. Which I guess is okay.

Yeah, I think that would be ok. We could also write the iteration in front of every logged message, so the user knows which is which. (Then we wouldn't necessarily need to reorder)

**Comment by cjmcgill** [comment]
> I see, then I guess we should keep the two different paths but if `--hyperopt_checkpoint_dir` is not specified we could set it to `--log_dir`

Yes I agree. And if `--log_dir` isn't specified, it should default to `--save_dir`.

**Comment by hesther** [comment]
> > Yeah, I think that would be ok. We could also write the iteration in front of every logged message, so the user knows which is which. (Then we wouldn't necessarily need to reorder)
> 
> There might be a way to identify unique instances of the hyperopt search. I'll try that in the next set of edits for testing.

Or simply use the seeds to identify each trial?

**Comment by cjmcgill** [comment]
> in sequential mode, once TPE enters into the focussed search, each new parameter set is chosen based on the previous results. E.g. if trials 1-10 already exist, the parameters for trial 11 are chosen based on trial 1-10. When running in parallel, every machine accesses the same previous checkpoints to come up with a new parameter set to try. So, machine 1 would conduct trial 11 based on trial 1-10 and machine 2 would conduct trial 12 based on trial 1-10 (not 1-11). Trial 11 and 12 will likely not have the same parameters because TPE has some randomness in how it chooses a new set (I think). So the sequential and parallel mode will naturally yield different outcomes. I think we should add one sentence to the README, saying that they are not fully equivalent. Furthermore, running in parallel leads to a larger number of iterations than specified in num_iters, because e.g. for num_iters=20 a new iteration can be started on multiple machines if 19 trials have finished, leading to num_iters+num_processes-1 trials (or less, depending on the timing of the individual processes). If we wanted to circumvent that, we would have to count the number of started processes in addition to the finished ones (but that is difficult to do).

Yes, the new instances are blind to parameter sets that are running but not yet finished. I tried to find a way to create new records the running instances could create to claim those parameters and optimize around them, but I was unable to do that in a way that hyperopt would respect. It might be possible with dummy results data but my initial tries there weren't successful. That would cause an additional problem of clearing them from the directory manually if you ever had an instance that failed for some reason. Thankfully the random seed seems to prevent duplicates, but that wouldn't be impossible especially in a small space.

Running parallel does cause something of a count issue. You could have each instance beyond the first accidentally run an extra random search iteration, making it extra important to log. You would also have as a nearly unavoidable behavior have n-1 extra total trials run. I can't figure out how to correct for this without accidentally counting failed jobs. I was figuring that extra iterations were less of a problem than accidentally having fewer iterations. What do you think?

Both of these warrant mention in the readme.

The most pressing problem related to this that I found was parallel instances not considering the extra end trials when logging the best instance, which is why I changed it to reload all checkpoints before reporting a best.

**Comment by cjmcgill** [comment]
@hesther I pushed a new commit that I think addresses all of our discussion points

File: chemprop/hyperopt_utils.py (hunk 1 above)
Comment (hesther) [reply chemprop/hyperopt_utils.py:64]: Ah, that is super interesting and makes sense, thank you!

File: chemprop/hyperparameter_optimization.py (hunk 2 above)
Comment (hesther) [reply chemprop/hyperparameter_optimization.py:138]: Haha ok well then I think we should just keep it as it is! Thanks for trying, though!

**Review by hesther — COMMENTED** [review]
Just tested all the functions again. I think you have addressed everything we talked about. Regarding the code itself, we still need to check the whitespaces, they are a bit inconsistent (will mark in code as I see them)

File: chemprop/hyperopt_utils.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    Checkpoints are newly loaded in at each iteration to allow for parallel entries
+    into the checkpoint folder by independent hyperoptimization instances.
+
+    :param dir_path: Path to the directory containing hyperopt checkpoint files.
+    :param previous_trials: Any previously generated trials objects that the loaded trials will be merged with.
+    :return: A trials object containing the merged trials from all checkpoint files.
+    """
+
+    # List out all the pickle files in the hyperopt checkpoint directory
+    hyperopt_checkpoint_files = [os.path.join(dir_path,path) for path in os.listdir(dir_path) if '.pkl' in path]
```
Comment (hesther) [inline chemprop/hyperopt_utils.py:54]: whitespaces for join(dir_path, path)

File: chemprop/hyperopt_utils.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+    # Load hyperopt trials object from each file
+    loaded_trials = Trials()
+    if previous_trials is not None:
+        loaded_trials = merge_trials(loaded_trials,previous_trials.trials)
```
Comment (hesther) [inline chemprop/hyperopt_utils.py:59]: whitespaces

File: chemprop/hyperopt_utils.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    else:
+        makedirs(seed_path, isfile=True)
+
+    seeds = [int(sd) for sd in seeds]
+
+    while seed in seeds:
+        seed += 1
+    seeds.append(seed)
+
+    write_line = " ".join(map(str,seeds)) + '\n'
```
Comment (hesther) [inline chemprop/hyperopt_utils.py:108]: whitespace

File: chemprop/hyperopt_utils.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+    with open(seed_path,'w') as f:
```
Comment (hesther) [inline chemprop/hyperopt_utils.py:110]: I think a whitespace (?)

File: chemprop/hyperopt_utils.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        'reaction',
+        'split_type',
+        'warmup_epochs',
+    ]
+
+    manual_trials_data = []
+    for i,trial_dir in enumerate(manual_trials_dirs):
+
+        # Extract trial data from test_scores.csv
+        with open(os.path.join(trial_dir,'test_scores.csv')) as f:
```
Comment (hesther) [inline chemprop/hyperopt_utils.py:148]: whitespace

File: chemprop/hyperopt_utils.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+            reader=csv.reader(f)
+            next(reader)
+            read_line=next(reader)
+        mean_score = float(read_line[1])
+        std_score = float(read_line[2])
+        loss = (1 if hyperopt_args.minimize_score else -1) * mean_score
+
+        # Extract argument data from args.json
+        with open(os.path.join(trial_dir,'args.json')) as f:
```
Comment (hesther) [inline chemprop/hyperopt_utils.py:157]: whitespace

**Comment by cjmcgill** [comment]
@hesther spaces added behind the commas you noted. One of these times I'll get that right the first time!

**Review by hesther — APPROVED** [review]
Looks good to me! (And lol I always take 2 iterations to get the whitespaces right ...)
